test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,211 @@
+      name: {}
+      connect_timeout: 5s
+      type: STATIC
+      load_assignment:
+        cluster_name: {}
+        endpoints:
+        - lb_endpoints:
+          - endpoint:
+              address:
+                socket_address:
+                  address: {}
+                  port_value: {}
+      lb_policy: ROUND_ROBIN
+      http2_protocol_options: {{}}
+    )EOF",
+                    name, name, Network::Test::getLoopbackAddressString(ipVersion()),
+                    fake_upstreams_[0]->localAddress()->ip()->port()));
+  }
+
+  void initializeCds() {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Setup cds and corresponding gRPC cluster.
+      auto* cds_config =
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+      bootstrap.mutable_dynamic_resources()->clear_ads_config();
+      bootstrap.mutable_dynamic_resources()->clear_lds_config();
+
+      cds_config->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+      cds_config->mutable_request_timeout()->set_seconds(1);
+      auto* grpc_service = cds_config->add_grpc_services();
+      setGrpcService(*grpc_service, "my_cds_cluster", fake_upstreams_[0]->localAddress());
+      grpc_service->mutable_envoy_grpc()->set_cluster_name("my_cds_cluster");
+
+      auto* cds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      cds_cluster->set_name("my_cds_cluster");
+      cds_cluster->mutable_connect_timeout()->set_seconds(5);
+      auto* sockaddr = cds_cluster->add_hosts()->mutable_socket_address();
+      sockaddr->set_protocol(envoy::api::v2::core::SocketAddress::TCP);
+      sockaddr->set_address(Network::Test::getLoopbackAddressString(ipVersion()));
+      sockaddr->set_port_value(0);
+      cds_cluster->clear_http2_protocol_options();
+    });
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+
+    HttpIntegrationTest::initialize();
+
+    fake_upstreams_[0]->set_allow_unexpected_disconnects(false);
+    // Causes cds_connection_ to be filled with a newly constructed FakeHttpConnection.
+    AssertionResult result =
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, cds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = cds_connection_->waitForNewStream(*dispatcher_, cds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    cds_stream_->startGrpcStream();
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}));
+    sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster,
+                                                   {buildCluster("cluster_0")}, "1");
+  }
+
+  testing::NiceMock<Server::Configuration::MockTransportSocketFactoryContext> factory_context_;
+  FakeStreamPtr cds_stream_;
+  FakeHttpConnectionPtr cds_connection_;
+  FakeHttpConnectionPtr upstream_connection_;
+};
+
+// GoogleGrpc causes problems.
```

Comment:
Do you know why? Maybe it just works now that you've debugged everything else? :)

Reply:
Wow, yes. I had tried it again fairly late in the process, I thought essentially immediately before submitting this PR, but I tried just now and it does work.